[service] Detect soft merge changes with a GitHub Action

prince-chrismc opened the pull request:
GitHub Config!

Following up from the feedback in #4725, I am pleased to say I was wrong... This is doable in an Action! [old concerns](https://github.com/conan-io/conan-center-index/pull/4725#issuecomment-788932688)

***

Things I am curious about... what do you build in Jenkins? Which commit do you choose? 

- Are you building based on Jenkins merges and branch index?
- The head of the pull request?

I have a sneaky suspicion it might be a relevant.

↪️ Every single push master will trigger an update on **all** open PRs
↪️ most changes to a PR will update _only_ that PR

### Demo

![image](https://user-images.githubusercontent.com/16867443/111055190-5e9cb100-8441-11eb-997c-339f05f72157.png)

prince-chrismc commented on .github/workflows/label_conflicts.yml:
```diff
@@ -0,0 +1,20 @@
+    steps:
+      - uses: prince-chrismc/label-merge-conflicts-action@v1
+        with:
+          conflict_label_name: "has conflicts"
```
A label needs to be selected/created

alternatively?

```suggestion
          conflict_label_name: "needs rebase"
```

mathbunnyru replied:
I'm against "needs rebase" - this will make people think they need to rebase, but after rebase it's impossible to review the PR in incremental mode.
`has conflicts` or `merge conflicts` work fine for me.

prince-chrismc replied:
That's a good point! what about `merge master`?

mathbunnyru replied:
Also works for me quite well.

prince-chrismc commented on .github/workflows/label_conflicts.yml:
```diff
@@ -0,0 +1,20 @@
+      - uses: prince-chrismc/label-merge-conflicts-action@v1
+        with:
+          conflict_label_name: "has conflicts"
+          github_token: ${{ secrets.GITHUB_TOKEN }}
```
You can use the @conan-center-bot's PAT if you'd like it to apply the labels

prince-chrismc replied:
https://github.com/prince-chrismc/label-merge-conflicts-action#faq---how-do-i-fix-resource-not-accessible-by-integration

This needs to be changes.

prince-chrismc commented:
This action fails to detect some corner cases

![image](https://user-images.githubusercontent.com/16867443/112637783-e5279a00-8e14-11eb-8ef1-4861ce6f794d.png)

prince-chrismc commented:
On March 1st 2021, there was changes to some of GitHub policies about Action permissions from Forks... I fixed a few security vulnerabilities and updated this PR accordingly.

jgsogo commented:
I know I sometimes play the role of devil's advocate, but someone needs to do it 😁 

I would like to add fine-grained permissions to these actions, specially those that are executing code outside our control (even if we really trust the user ❤️ ). I've read about permissions [here](https://docs.github.com/en/actions/reference/authentication-in-a-workflow#permissions-for-the-github_token) and [here](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#permissions). I would say we need to read and write (add the label) for pull-requests. Probably we need to change the default at conan-io level to be restrictive and declare explicitly in the job what it needs.

prince-chrismc commented:
The second link is new 😱 I must have missed a new blog post! [this is it](https://github.blog/changelog/2021-04-20-github-actions-control-permissions-for-github_token/). I'll make some time in the next few days and see what it means on my end 🛩️ 

I also noticed [this new feature](https://github.blog/changelog/2021-04-22-github-actions-maintainers-must-approve-first-time-contributor-workflow-runs/) on another project I contribute to (It requires contributors with write access to approve actions)
![image](https://user-images.githubusercontent.com/16867443/118148503-a064bc00-b3de-11eb-9bcb-1a77cb27d127.png)

[GitHub killing passwords](https://github.blog/2021-05-10-security-keys-supported-ssh-git-operations/)

jgsogo commented:
Yep, in another project, I found quite inconvenient the change they did on March 1st 2021. I understand why and I think it is needed, but my action was no longer able to post a message to the PR (it was just a random gif 😄 ), and the workaround was a bit complex.

Now I'm surprised also about all this new stuff, they are taking permissions seriously... and I also like the first-time contributors approval. Please, share your thoughts here. Thanks, and sorry for so much back and forth...

prince-chrismc commented:
So I found the minimum set of per missions (sorry for the fails runs) which make sense.

I think for trusting the code running an option could be pinning the exact git SHA instead of the moving major release.

prince-chrismc commented:
After more testing I found some short commings, which I've adjusted. prince-chrismc/conan-center-index#27 is working after my more recent changes.

I also had some success with https://docs.github.com/en/github/administering-a-repository/managing-repository-settings/disabling-or-limiting-github-actions-for-a-repository#allowing-specific-actions-to-run
in my settings I only allow my own action in order to prevent someone from editing this workflow and adding malicious code (too easily)

prince-chrismc commented:
I have finally add the "comment when conflicts are detected" with a polite message for OP if this is merged
